### PR TITLE
fix: large snapshot diff recursion error

### LINKED ISSUE #577: RecursionError raised by _fancy_replace for a large snapshot
**Describe the bug**

_fancy_replace will encounter an error when analyzing a string that is large, repetitive, and slightly different from the expected snapshot.

**To reproduce**
1. Create this test:
```
CHANGED = False
def test_big_snapshot(
    snapshot,
):
    if CHANGED:
        result = [str(x) + ("a" * 20) for x in range(3000)]
    else:
        result = [str(x) + ("a" * 20 + "b") for x in range(3000)]

    assert result == snapshot
```
2. Run --snapshot-update with CHANGED = False
3. Run tests with CHANGED = True

```
(repeats)
../lib/python3.8/difflib.py:1015: in _fancy_replace
    yield from self._fancy_helper(a, best_i+1, ahi, b, best_j+1, bhi)
../lib/python3.8/difflib.py:1027: in _fancy_helper
    yield from g
../lib/python3.8/difflib.py:1015: in _fancy_replace
    yield from self._fancy_helper(a, best_i+1, ahi, b, best_j+1, bhi)
../lib/python3.8/difflib.py:1027: in _fancy_helper
    yield from g
../lib/python3.8/difflib.py:1015: in _fancy_replace
    yield from self._fancy_helper(a, best_i+1, ahi, b, best_j+1, bhi)
../lib/python3.8/difflib.py:1027: in _fancy_helper
    yield from g
../lib/python3.8/difflib.py:969: in _fancy_replace
    cruncher.ratio() > best_ratio:
../lib/python3.8/difflib.py:644: in ratio
    matches = sum(triple[-1] for triple in self.get_matching_blocks())
../lib/python3.8/difflib.py:479: in get_matching_blocks
    i, j, k = x = self.find_longest_match(alo, ahi, blo, bhi)
../lib/python3.8/difflib.py:444: in find_longest_match
    return Match(besti, bestj, bestsize)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

_cls = <class 'difflib.Match'>, a = 0, b = 0, size = 26

>   ???
E   RecursionError: maximum recursion depth exceeded while calling a Python object

```
**Expected behavior**

No error should be raised and the test should fail with a reasonable message. It would be fine to not have the full fancy error message for large snapshots.

**Environment (please complete the following information):**

- OS: Windows
- Syrupy Version: 1.5.0
- Python Version: 3.8

**Additional context**

The real world use case was a 3000-row DataFrame converted to csv which had one column that was the same string for every row. When the string changed by 1 character the test appeared to hang (took more than an hour).

### PR DESCRIPTION
**Pull request by iamogbz**
## Description

Avoids the recursion error limit in `difflib.ndiff` by truncating the checked lines to a subsection of the diff.

* Documentation update and formatting
* Poetry lock update via `inv install`

## Related Issues

- Fixes #577

- Helps #581, #625

## Checklist

- [x] This PR has sufficient documentation.
- [x] This PR has sufficient test coverage.
- [x] This PR title satisfies semantic [convention](https://www.conventionalcommits.org/en/v1.0.0/#summary).

## Additional Comments

Went with the simple approach of just checking from the first mismatch and giving a large enough buffer to catch meaningful differences.

Can be validated by setting the `DIFF_LINE_COUNT_LIMIT` to a high value e.g. `100000` and running tests, maybe a follow up can make this configurable;

```sh
(syrupy-py3.11) ➜  syrupy git:(fix-large-snapshot-recursion-error) ✗ inv test -t test_diff_large
====================================== test session starts ======================================
platform darwin -- Python 3.11.4, pytest-7.4.0, pluggy-1.0.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/Emmanuel/Sources/github/tophat/syrupy
configfile: pyproject.toml
plugins: syrupy-4.0.6, xdist-3.3.1, benchmark-4.0.0
collected 256 items / 254 deselected / 2 selected                                               

tests/syrupy/extensions/test_base.py FF                                                   [100%]

=========================================== FAILURES ============================================
____________________ TestSnapshotReporter.test_diff_large[SnapshotReporter] _____________________

self = <tests.syrupy.extensions.test_base.TestSnapshotReporter object at 0x1067e0c10>
Reporter = <class 'syrupy.extensions.base.SnapshotReporter'>
osenv = <function env_context at 0x105db7a60>

    def test_diff_large(self, Reporter, osenv):
        n_count = 3000
        obj_a = [str(x) + ("a" * 20) for x in range(n_count)]
        obj_b = [line_a + "b" for line_a in obj_a]
        str_a = "\n".join(obj_a)
        str_b = "\n".join(obj_b)
        with osenv(NO_COLOR="true"):
>           assert (
                len(list(Reporter().diff_lines(str_a, str_b)))
                <= DIFF_LINE_COUNT_LIMIT * 2
            )

tests/syrupy/extensions/test_base.py:44: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/syrupy/extensions/base.py:271: in diff_lines
    for line in self.__diff_lines(str(snapshot_data), str(serialized_data)):
src/syrupy/extensions/base.py:295: in __diff_lines
    for line in self.__diffed_lines(a, b):
src/syrupy/extensions/base.py:314: in __diffed_lines
    for line in qdiff(a.splitlines(keepends=True), b.splitlines(keepends=True)):
../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:872: in compare
    yield from g
...
../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:997: in _fancy_helper
    yield from g
../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:985: in _fancy_replace
    yield from self._fancy_helper(a, best_i+1, ahi, b, best_j+1, bhi)
../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:997: in _fancy_helper
    yield from g
../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:915: in _fancy_replace
    cruncher = SequenceMatcher(self.charjunk)
../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:182: in __init__
    self.set_seqs(a, b)
../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:194: in set_seqs
    self.set_seq2(b)
../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:248: in set_seq2
    self.__chain_b()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <difflib.SequenceMatcher object at 0x107a71350>

    def __chain_b(self):
        # Because isjunk is a user-defined (not C) function, and we test
        # for junk a LOT, it's important to minimize the number of calls.
        # Before the tricks described here, __chain_b was by far the most
        # time-consuming routine in the whole module!  If anyone sees
        # Jim Roskind, thank him again for profile.py -- I never would
        # have guessed that.
        # The first trick is to build b2j ignoring the possibility
        # of junk.  I.e., we don't call isjunk at all yet.  Throwing
        # out the junk later is much cheaper than building b2j "right"
        # from the start.
        b = self.b
        self.b2j = b2j = {}
    
        for i, elt in enumerate(b):
            indices = b2j.setdefault(elt, [])
            indices.append(i)
    
        # Purge junk elements
        self.bjunk = junk = set()
        isjunk = self.isjunk
        if isjunk:
>           for elt in b2j.keys():
E           RecursionError: maximum recursion depth exceeded while calling a Python object

../../../../.pyenv/versions/3.11.4/lib/python3.11/difflib.py:288: RecursionError
==================================== short test summary info ====================================
FAILED tests/syrupy/extensions/test_base.py::TestSnapshotReporter::test_diff_large[SnapshotReporter] - RecursionError: maximum recursion depth exceeded while calling a Python object
FAILED tests/syrupy/extensions/test_base.py::TestSnapshotReporter::test_diff_large[SnapshotReporterNoContext] - RecursionError: maximum recursion depth exceeded while calling a Python object
========================= 2 failed, 254 deselected in 136.72s (0:02:16) =========================
```

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -19,7 +19,7 @@
 	"settings": {
 		"python.defaultInterpreterPath": "/usr/local/bin/python",
 		"python.linting.enabled": true,
-		"python.linting.pylintEnabled": true,
+		"python.linting.pylintEnabled": false,
```
Comment (iamogbz) [inline .devcontainer/devcontainer.json:22]: pylint not used in project

File: tests/syrupy/extensions/amber/test_amber_snapshot_diff.py [hunk 2]
```diff
@@ -27,28 +27,28 @@ def test_snapshot_diff_id(snapshot):
         "field_4": 1,
         "field_5": False,
         "field_6": (True, "hey", 2, None),
-        "field_7": {False, "no", 0, None},  # noqa: disable=B033
```
Comment (iamogbz) [inline tests/syrupy/extensions/amber/test_amber_snapshot_diff.py:30]: This was adding no value just confusion

File: src/syrupy/extensions/amber/serializer.py [hunk 3]
```diff
@@ -113,7 +113,8 @@ def write_file(
         with open(filepath, "w", encoding=TEXT_ENCODING, newline=None) as f:
             f.write(f"{cls._marker_prefix}{cls.Marker.Version}: {cls.VERSION}\n")
             for snapshot in sorted(
-                snapshot_collection, key=lambda s: cls._snapshot_sort_key(s)  # type: ignore # noqa: E501
+                snapshot_collection,
+                key=cls._snapshot_sort_key,  # noqa: E501
```
Comment (iamogbz) [inline src/syrupy/extensions/amber/serializer.py:117]: Linter flagged as redundant

File: tests/syrupy/extensions/amber/test_amber_snapshot_diff.py [hunk 4]
```diff
@@ -27,28 +27,28 @@ def test_snapshot_diff_id(snapshot):
...
+        "field_7": {False, "no", None},
```
Comment (iamogbz) [inline tests/syrupy/extensions/amber/test_amber_snapshot_diff.py:30]: This was adding no value just confusion

**Review by noahnu — APPROVED** [review]

**Pull request merged**